doc/contributing.rst

```diff
@@ -412,12 +412,114 @@ To avoid accidental highlighting follow this rule in ``*.nim`` files:
+Reasoning
+<<<<<<<<<
+
+One of the key goals of the |nimskull| project is sustainability. Writing
+good commit messages is important, since future contributors need to have a
+way to reason about decitions made in the past.
+
+It should be clear as to *"what"* change was made. Additional details like
+high-level oviewview of *"how"* and *"why"* are also welcome, but should go
+into commit body. Of course *"what"* can be contiued in the body as well.
+
+Rules
+<<<<<
+
+Commit message should contain enough information in order for end user to
+understand what is going on by looking at the github log.
+
+Example of bad commit message::
+
+  Fixes #123; refs #124
+
+indicates that issue ``#123`` is completely fixed (GitHub may automatically
+close it when the PR is committed), wheres issue ``#124`` is referenced
+(e.g.: partially fixed) and won't close the issue when committed. This
+tells reader **nothing** about what the commit did, requires going into two
+separate github pages and reading through multiple comments in order to
+figure out what was *wrong* - not what *changed* in the commit.
+
+Instead, commit title should contain a human-readable text (we are
+optimizing for humans after all)
+
+- Keep the title length to 50 characters, and commit message body to 72
+  characters.
+- If you want to use automatic "fixes" feature of the github, please supply
+  full URL in the body instead, it makes it much easier to copy-paste and
+  see (with pure hash you first need to understand whether change was in
+  ``issues/`` or ``pull/``, then open random PR/issue and replace hash
+  URL).
+- Consider using "Fixes" for commits that fix bugs, and "Closes" for
+  commits that implement features.
+- Most of the points on this list were derived from `How To Write a Good
+  Commit Message
+  <https://api.coala.io/en/latest/Developers/Writing_Good_Commits.html>`_ -
+  we suggest reading it as well.
+
+Crafting good commit title might be hard at first. Some simple rules to
+look out for:
+
+- Avoid using phrases like ``"minor fixes for XXX"``, ``"various
+  improvements"``, ``"tiny patch"`` - they don't add any substantial
+  information and only clutter the title.
+- No bare github hashes in the title. Things like ``fix #19193 (#19195)
+  [backport:1.2]`` are absolutely illegible.
+
+When writing commit message consider who are you writing this message
+for:
+
+- you or someone else in the far future found a big and stumbled upon
+  this in a git bisect
+- you or someone else is going through recent commits trying to write a
+  changelog
+- you or someone else is trying to find a bigt that can't be replicated
+  reliably so they're reading commits seeing which might be related
+- someone new to the code base is trying to learn bits and pieces about
+  what's being worked on and what types of things are changing
+- you or someone else is sorting out documentation or test gaps, or
+  seeing if old issues can be closed with recent changes
+
+
+Message examples
```

Comment:
Specific examples are up to debate, and I think having more would be beneficial. For now I added a simple example, in future this will be expanded